HULKs/hulk
# Ingamepenaltykick

## Introduced Changes
Added behaviors to handle in game penalty kick:
When defending:
Goal keeper has position restrictions
Defenders have an offset
Striker should not try and dribble
No players may enter penalty box
When attacking:
Striker assumes position to take penalty shot
No other players may enter penalty box.

To this end behaviors were added, ball state and game state were adjusted, penalty box obstacle creating for path planning was added.

Fixes #137 

## ToDo / Known 

~~Undefined behavior if Striker is outside penalty box when preparing to execute penalty kick as penalty box obstacles would prevent him from entering the penalty box. Undefined as this would only happen if striker is de-localized in ready phase but re-localizes in set or at beginning of playing phase so that he positions himself wrong in ready phase and cannot correct during playing/executing the penalty kick. Must be observed in real games.~~
Striker ignores penalty box when Hulks are kicking team. Penalty kick location should be changed to in order of complexity and iteration:
1. better distance
2. Isometry for angling shots
3. Kick decision based location

~~defend_*_pose and defend_penalty_*_pose could be merged and position_to_defend decided based on game controller state but removing the lines that set the specific ball position for the penalty case seems to break the behavior~~

## Ideas for Next Iterations (Not This PR)

Add option for goal keeper to jump to block goal.

~~Many improvements could be made as this currently has some duplicate but necessary code. This is essentially the precursor to several probably necessary improvements to the behavior architecture. ~~
Important would be to move all instances of evaluating the sub state to one file (game_state_filter?) so that it is not repeatedly evaluated. Also an improved system of action to role assignments in this type of edge case would be convenient.

## How to Test

Test game needed

The lua files ingamepenalty_* provide various scenarios to test the new behavior, final testing will have to be done in a test game to ensure functionality of some components not implemented in behavior simulator (i.e. verifying correctness of head motion commands).
